bot.py

```diff
@@ -114,42 +115,43 @@ def start_method (self):
     def main_loop (self):
         while True:
             if self._is_ringing:
-                if time.time () > self._rhythm.index_to_real_time (self._row_number, self._place):
-                    bell = self._place if self._is_ringing_rounds else self._row [self._place]
-
-                    if not self._tower.user_controlled (bell):
-                        self._tower.ring_bell (bell, self.is_handstroke)
-
-                    self._place += 1
-
-                    if self._place == self.stage:
-                        self._row_number += 1
-                        self._place = 0
-
-                        if not self._is_ringing_rounds:
-                            self.start_next_row ()
-                        else:
-                            for b in range (self.stage):
-                                self.expect_bell (b, b)
-
-                        if self._is_ringing_rounds:
-                            if self._row_number == 2 and self.do_up_down_in:
-                                self._should_start_method = True
-                        
-                        if self._row_number % 2 == 0:
-                            # We're just starting a handstroke
-                            if self._should_stand:
-                                self._should_stand = False
-                                self._is_ringing = False
-
-                            if self._should_start_method and self._is_ringing_rounds:
-                                self._should_start_method = False
-                                self._is_ringing_rounds = False
-
-                                self.start_method ()
-
-                            if self._should_start_ringing_rounds and not self._is_ringing_rounds:
-                                self._should_start_ringing_rounds = False
-                                self._is_ringing_rounds = True
+                bell = self._place if self._is_ringing_rounds else self._row [self._place]
+
+                user_controlled = self._tower.user_controlled(bell)
+                self._rhythm.wait_for_bell_time(time.time(), bell, self._row_number, self._place, user_controlled, self.is_handstroke)
```

Comment:
This change just pushes the sleeping down into `self._rhythm.wait_for_bell_time` instead of polling it every 0.01s

That then allows the `WaitForUserRhythm` to spin or a `Rhythm` to calculate a single sleep

Everything below is then just indentation